README
Web Scraping with Python
------------------------


Kelley Blue Book
----------------
This was my first serious project that I uploaded to Github. It has been moved to this directory to make my projects look a little bit cleaner. I used BeautifulSoup to scrape KBB and then cleaned the data. I thought that I was going to be able to do more with it than what I was. I think I will come back to this at some point.


2019 Movie Scraping from Wikipedia
----------------------------------
This is a project where I scraped and cleaned data from a Wikipedia table containing information about 2019 films.  

I then did some graphing of the data and attempted to perform linear regression on the data with PyTorch where I was using the budget to predict the box office performance.  The R2 score was 64%, however the loss was extremely high.


Linux Distributions
-------------------
I scraped the top 275 distributions on https://distrowatch.com/ and got the basic specs (e.g., distribution it is based on, architecture, type of desktop) and the community rating for each of them. Then I created a dataframe that has the features of every version of the distribution (there is a .csv containing this). For this project I only used the most recent release.

I 'one-hot-encoded' the data and then classified it based on its' rating to perform a logistic regression. The model did better than chance, however the data that I used may not be the best.  This was just a little project I did because I just recently got into Linux. 

Overall, I would have to say that (with the data I have) the specs of a distribution does not really correlate with what people think of it.


Trulia House Prices
-------------------
Scraped 2500 house prices from 5 cities in Missouri. I then cleaned the data and created a few graphs.  The R2 score from the linear regression performed was 0.72, however the MSE was huge.  Still a work in progress.


CHANGELOG.md
### modified email (HEAD -> master)
>Mon, 7 Jun 2021 21:33:52 -0500

>Author: Lucas Burns (lmb@lmburns.com)

>Commiter: Lucas Burns (lmb@lmburns.com)




### domain switch (origin/master, home/master, gitlab/master)
>Thu, 3 Jun 2021 13:11:21 -0500

>Author: Lucas Burns (lucas@burnsac.xyz)

>Commiter: Lucas Burns (lucas@burnsac.xyz)




### functionized the rest, created dictionaries for mappings
>Wed, 2 Jun 2021 13:34:32 -0500

>Author: Lucas Burns (lucas@burnsac.xyz)

>Commiter: Lucas Burns (lucas@burnsac.xyz)




### using sspipe to do unix style pipes - more readable
>Mon, 31 May 2021 12:18:55 -0500

>Author: Lucas Burns (lucas@lucasburns.xyz)

>Commiter: Lucas Burns (lucas@lucasburns.xyz)




### started functionizing
>Mon, 31 May 2021 01:13:06 -0500

>Author: Lucas Burns (lucas@lucasburns.xyz)

>Commiter: Lucas Burns (lucas@lucasburns.xyz)




### forgot to commit earlier
>Wed, 26 May 2021 15:06:39 -0500

>Author: Lucas Burns (lucas@lucasburns.xyz)

>Commiter: Lucas Burns (lucas@lucasburns.xyz)




### added 4chan scraper - initial commit for this project
>Sat, 22 May 2021 09:34:15 -0500

>Author: Lucas Burns (burnsac@me.com)

>Commiter: Lucas Burns (burnsac@me.com)




### fixed grammar readme
>Mon, 22 Mar 2021 10:45:08 -0500

>Author: Lucas Burns (burnsac@me.com)

>Commiter: Lucas Burns (burnsac@me.com)




### modified line in readme
>Sun, 21 Mar 2021 22:07:07 -0500

>Author: Lucas Burns (burnsac@me.com)

>Commiter: Lucas Burns (burnsac@me.com)




### updated slashdot slightly
>Wed, 10 Mar 2021 13:01:45 -0600

>Author: Lucas Burns (burnsac@me.com)

>Commiter: Lucas Burns (burnsac@me.com)




### updated readme, added youtube.py
>Wed, 3 Mar 2021 23:33:29 -0600

>Author: Lucas Burns (burnsac@me.com)

>Commiter: Lucas Burns (burnsac@me.com)




### furthing slashdot
>Wed, 3 Mar 2021 10:52:22 -0600

>Author: Lucas Burns (burnsac@me.com)

>Commiter: Lucas Burns (burnsac@me.com)




### added dataframe & testing stagit recompile
>Sun, 28 Feb 2021 19:32:01 -0600

>Author: Lucas Burns (burnsac@me.com)

>Commiter: Lucas Burns (burnsac@me.com)




### updated
>Sun, 28 Feb 2021 18:47:03 -0600

>Author: Lucas Burns (burnsac@me.com)

>Commiter: Lucas Burns (burnsac@me.com)




### -
>Sun, 7 Feb 2021 20:26:28 -0600

>Author: Lucas Burns (burnsac@me.com)

>Commiter: Lucas Burns (burnsac@me.com)




### version control py
>Fri, 5 Feb 2021 14:47:38 -0600

>Author: Lucas Burns (burnsac@me.com)

>Commiter: Lucas Burns (burnsac@me.com)




### version control py
>Fri, 5 Feb 2021 14:41:46 -0600

>Author: Lucas Burns (burnsac@me.com)

>Commiter: Lucas Burns (burnsac@me.com)




### modified readme (homemaster)
>Tue, 12 Jan 2021 14:59:46 -0600

>Author: Lucas Burns (burnsac@me.com)

>Commiter: Lucas Burns (burnsac@me.com)




### added yt-transcripts
>Mon, 11 Jan 2021 10:07:29 -0600

>Author: Lucas Burns (burnsac@me.com)

>Commiter: Lucas Burns (burnsac@me.com)




### modified README > compatibility for plain txt
>Thu, 7 Jan 2021 16:10:29 -0600

>Author: burnsac5040 (burnsac@me.com)

>Commiter: burnsac5040 (burnsac@me.com)




### modified README > txt
>Thu, 7 Jan 2021 16:05:40 -0600

>Author: burnsac5040 (burnsac@me.com)

>Commiter: burnsac5040 (burnsac@me.com)




### removed In and blank lines python
>Thu, 7 Jan 2021 09:21:55 -0600

>Author: burnsac5040 (burnsac@me.com)

>Commiter: burnsac5040 (burnsac@me.com)




### testing server (gitwebmaster)
>Mon, 4 Jan 2021 12:41:09 -0600

>Author: burnsac5040 (burnsac@me.com)

>Commiter: burnsac5040 (burnsac@me.com)




### testing server
>Mon, 4 Jan 2021 00:15:45 -0600

>Author: burnsac5040 (burnsac@me.com)

>Commiter: burnsac5040 (burnsac@me.com)




### testing server
>Mon, 4 Jan 2021 00:13:41 -0600

>Author: burnsac5040 (burnsac@me.com)

>Commiter: burnsac5040 (burnsac@me.com)




### trulia
>Sun, 3 Jan 2021 18:58:33 -0600

>Author: burnsac5040 (burnsac@me.com)

>Commiter: burnsac5040 (burnsac@me.com)




### added python files
>Sun, 3 Jan 2021 18:10:11 -0600

>Author: burnsac5040 (burnsac@me.com)

>Commiter: burnsac5040 (burnsac@me.com)




### added python files
>Sun, 3 Jan 2021 18:08:55 -0600

>Author: burnsac5040 (burnsac@me.com)

>Commiter: burnsac5040 (burnsac@me.com)




### edited readme
>Tue, 29 Dec 2020 14:30:38 -0600

>Author: burnsac5040 (burnsac@me.com)

>Commiter: burnsac5040 (burnsac@me.com)




### modified logreg, added href2distro & wiki
>Mon, 21 Dec 2020 09:59:21 -0600

>Author: burnsac5040 (burnsac@me.com)

>Commiter: burnsac5040 (burnsac@me.com)




### added kbb
>Sat, 19 Dec 2020 23:48:51 -0600

>Author: burnsac5040 (burnsac@me.com)

>Commiter: burnsac5040 (burnsac@me.com)




### added distrowatch / reorganized repo
>Sat, 19 Dec 2020 19:28:31 -0600

>Author: burnsac5040 (burnsac@me.com)

>Commiter: burnsac5040 (burnsac@me.com)






trulia/trulia_scraper.py
# ---
# jupyter:
#   jupytext:
#     formats: ipynb,py
#     text_representation:
#       extension: .py
#       format_name: light
#       format_version: '1.5'
#       jupytext_version: 1.9.1
#   kernelspec:
#     display_name: Python 3
#     language: python
#     name: python3
# ---

# +
from bs4 import BeautifulSoup as bs
import requests

import numpy as np 
import pandas as pd 
import time
import ssl
import re
import csv
import datetime

# +
base_url = 'https://www.trulia.com'
url = 'https://www.trulia.com/for_sale/Columbia,MO/1p_beds/SINGLE-FAMILY_HOME_type/'
headers = {
    'accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9',
    'accept-encoding': 'gzip, deflate, br',
    'accept-language': 'en-US,en;q=0.9,es;q=0.8',
    #'cache-control': 'max-age=0',
    'dnt': '1',
    'sec-fetch-dest': 'document',
    'sec-fetch-mode': 'navigate',
    'sec-fetch-site': 'cross-site',
    'sec-fetch-user': '?1',
    'sec-gpc': '1',
    'upgrade-insecure-requests': '1',
    'user-agent':'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/87.0.4280.101 Safari/537.36'
    }

ctx = ssl.create_default_context()
ctx.check_hostname = False
ctx.verify_mode = ssl.CERT_NONE

r = requests.get(url, headers=headers, verify=False)
soup = bs(r.text, 'html.parser')


# -

# ## Single Search Page

# +
def get_page(url):
    """Returns a beautiful soup object."""
    r = requests.get(url, headers=headers, verify=False)
    
    if not r.ok:
        print('Server Responded: ', r.status_code)
    else:
        soup = bs(r.text, 'html.parser')
    return soup


def get_main_attrs(soup):
    """Gets the attributes of each listing from a search page and returns
    a dictionary where the address is the key and the attributes are the values."""
    addr_ = soup.find_all('div',attrs={'data-testid': 'property-street'})
    addr = [x.get_text().strip() for x in addr_]
    region_ = soup.find_all('div',attrs={'data-testid': 'property-region'})
    region = [x.get_text().strip() for x in region_]
    new_ = soup.find_all('div', attrs={'data-testid': 'property-tags'})
    new = [x.get_text().strip() if x.get_text() is not '' else 'OLD' for x in new_]
    prices_ = soup.find_all('div', attrs={'data-testid': 'property-price'})
    prices = [x.get_text().strip() for x in prices_]
    bedrooms_ = soup.find_all('div', attrs={'data-testid': 'property-beds'})
    bedrooms = [x.get_text().strip() for x in bedrooms_]
    baths_ = soup.find_all('div', attrs={'data-testid': 'property-baths'})
    baths = [x.get_text().strip() for x in baths_]
    sqft_ = soup.find_all('div', attrs={'data-testid': 'property-floorSpace'})
    sqft = [x.get_text().strip() for x in sqft_]

    c = zip(addr, region, new, prices, bedrooms, baths, sqft)
    d = {}

    for a, *x in c:
        d.setdefault(a, []).append(x)

    for k, v in d.items():
        d[k] = v[0]

    df = pd.DataFrame(d).T.rename(columns={0:'region', 1:'new', 2:'price', 3:'bedrm', 4:'bth', 5:'sqft'})
    return df


def get_urls(soup):
    """Gets the external URLs on the page."""
    hrefs = []

    for listing in soup.find_all('div', attrs={'data-testid': 'home-card-sale'}):
        for link in listing.find_all('a', attrs={'href': re.compile('^/')}):
            hrefs.append(link)
        
    return [base_url + x['href'] for x in hrefs]


# -

# ## Single Listing

# +
s_url = 'https://www.trulia.com/p/mo/columbia/2815-wild-plum-ct-columbia-mo-65201--2060813753'

page = get_page(s_url)


# -

def get_page_attrs(soup):
    """Gets a single page's attributes and returns a dataframe."""
    # Getting address again to map it with the address collected on main page
    try:
        addr_p = soup.find('span', attrs={'data-testid': 'home-details-summary-headline'}).get_text()
    except:
        addr_p = ''

    # Crime
    try:
        crime = soup.find('div',attrs={'aria-label': 'Crime'}).get_text()
    except:
        crime = ''
    # Schools in the area
    try:
        schools = soup.find('div', attrs={'aria-label': 'Schools'}).get_text()
    except:
        schools = ''

    # Home Details -- (heating, roof, etc.)
    try:
        details_ = soup.find('ul', attrs={'data-testid': 'home-features'})
        details = [x.get_text() for x in details_.find_all('li')]
    except:
        details = ''

    # History of listings
    try:
        list_hist_ = soup.find('div', attrs={'data-testid': 'price-history-container'}).find('table').find_all('tr')
        list_hist = [x.get_text().strip() for x in list_hist_]
    except:
        list_hist = ''

    # Taxes on the house
    try:
        tax_table_ = soup.find('div', attrs={'data-testid': 'property-tax-container'}).find('table').find_all('tr')
        tax_table = [x.get_text().strip() for x in tax_table_]
    except:
        tax_table = ''

    try:
        price_trends_ = soup.find_all('div', attrs={'class': 'Text__TextBase-sc-1i9uasc-0-div Text__TextContainerBase-sc-1i9uasc-1 epkfvN'})
        # Typical home value compared to others like it
        typ_home_val = [x.get_text() for x in price_trends_][-2] 
        # Typical price per sqft of houses similar to it
        typ_sqft_pri = [x.get_text() for x in price_trends_][-1]
    except:
        typ_home_val = ''
        typ_sqft_pri = ''

    try:
        pcts_ = soup.find_all('span', attrs={'class': 'Text__TextBase-sc-1i9uasc-0 fUDZSu'})
        # How much this house varies from others (based on typ_home_val)
        val_pct = [x.get_text() for x in pcts_][-2]
        # How much this house varies from others (based on typ_sqft_pri)
        sqft_pct = [x.get_text() for x in pcts_][-1]
    except:
        val_pct = ''
        sqft_pct = ''

    l = [addr_p, crime, schools, details, list_hist, tax_table, typ_home_val, val_pct, typ_sqft_pri, sqft_pct]

    names = ['addr', 'crime', 'schools', 'details', 'list_hist', 'tax', 'typ_val', 'val_pct', 'typ_sqft', 'sqft_pct']

    df = pd.DataFrame(l).T

    return df.rename(columns=dict(zip(df.columns.values, names)))


# ## Using a Single Search page to Scrape 30 Listings

# +
spage = 'https://www.trulia.com/for_sale/Columbia,MO/1p_beds/SINGLE-FAMILY_HOME_type/2_p/'

l_urls = get_urls(get_page(spage))

# +
dfs = []

for i, url in enumerate(l_urls):
    df = get_page_attrs(get_page(url))
    dfs.append(df)
    print(f'Listing {i:-<5} completed')

# +
df_main = get_main_attrs(get_page(spage)).rename_axis('addr').dropna(how='all')
df_ind = pd.concat(dfs).set_index('addr').dropna(how='all')

# df_main.join(df_ind, how='outer').dropna(how='all')
pd.merge(df_main, df_ind, left_index=True, right_index=True).head(1)
# -

# ## Getting Main Attributes for Multiple Pages
# - I got the ranges to use on the website by viewing the number of pages available

# +
from itertools import chain

como = [f'https://www.trulia.com/for_sale/Columbia,MO/1p_beds/SINGLE-FAMILY_HOME_type/{x}_p/' for x in range(1, 8)]
kc = [f'https://www.trulia.com/for_sale/Kansas_City,MO/1p_beds/SINGLE-FAMILY_HOME_type/{x}_p/' for x in range(1, 29)]
stl = [f'https://www.trulia.com/for_sale/Saint_Louis,MO/1p_beds/SINGLE-FAMILY_HOME_type/{x}_p/' for x in range(1, 38)]
spring = [f'https://www.trulia.com/for_sale/Springfield,MO/1p_beds/SINGLE-FAMILY_HOME_type/{x}_p/' for x in range(1, 15)]
lees = [f'https://www.trulia.com/for_sale/Lees_Summit,MO/1p_beds/SINGLE-FAMILY_HOME_type/{x}_p/' for x in range(1, 16)]

list_all_urls = [como, kc, stl, spring, lees]
all_urls = list(chain.from_iterable(list_all_urls))

len(all_urls)

# +
all_main = []

for i, url in enumerate(all_urls):
    df = get_main_attrs(get_page(url))
    all_main.append(df)
    if (i+1) % 5 == 0:
        print(f'URL {i+1:-<5} completed')

# -

df_main = pd.concat(all_main)
df_main.sample(5)
# df_main.to_csv('df_main.csv', index=True, columns=df_main.columns.values)

df_main = pd.read_csv('df/df_main.csv', index_col=0)

# ## Getting all of the individual listing's URLs
# - I had to separate the two for loops (main page attributes/individual page attributes) because I kept getting blocked/captcha.

# +
ind_urls = []

for idx, url in enumerate(all_urls):
    urls = get_urls(get_page(url))
    ind_urls.append(urls)
    time.sleep(3)
    if (idx+1) % 20 == 0:
        print(f'URL {idx+1:-<5} completed {urls[0][12:40]:-<5}')

# +
import pickle

page_urls = list(chain.from_iterable(ind_urls))

with open('pickle/page_urls.pickle', 'wb') as f:
    pickle.dump(page_urls, f, protocol=pickle.HIGHEST_PROTOCOL)

len(page_urls)

# +
ind_dfs = []

for idx, url in enumerate(page_urls):
    df = get_page_attrs(get_page(url))
    ind_dfs.append(df)
    time.sleep(2)

    if (idx+1) % 50 == 0:
        print(f'URL {idx+1:-<5}completed{url[28:40]:->15}')

# +
# Saving the list of dataframes
with open('pickle/df_ind.pickle', 'wb') as f:
    pickle.dump(ind_dfs, f, protocol=pickle.HIGHEST_PROTOCOL)

with open('pickle/df_ind.pickle', 'rb') as f:
    ind_list = pickle.load(f)

# Saving the concatenated dataframes
df_ind = pd.concat(ind_dfs).replace('', np.nan).dropna(how='all').set_index('addr')
df_ind.to_csv('df/df_ind.csv', index=True, columns=df_ind.columns.values)
# -

print(f'df_main: {df_main.shape}, df_ind: {df_ind.shape}')
